### PR TITLE
Add custom hooks and "password-migrate" handler

### PR DESCRIPTION
**Pull request by tyrw**
Deep
Closes Dev-270
Closes Dev-316

- Refactors all API login response handling into `handleLoginResponse`
  - Refactors authentication tests to mock `handleLoginResponse` rather than `Userfront.user` and `window` behavior
  - Adds tests to assert that `handleLoginResponse` properly affects `Userfront.user` and `window` behavior
- Moves MFA behavior into own `mfa.js` file (out of `authentication.js`)
- Adds custom handler methods
  - `handleUpstreamResponse` is run when `upstreamResponse` is included in the response
  - `handleMfaRequired` is run when "MFA required" response is received
  - `handlePkceRequired` is run when "PKCE required" response is received
  - `handleTokens` is run when `tokens` is included in the response
  - `handleRedirect` is always run if present, unless one of the above interrupts flow
- Renames default methods to match custom handler method names
  - `defaultHandleUpstreamResponse`
  - `defaultHandleMfaRequired`
  - `defaultHandlePkceRequired`
  - `defaultHandleTokens`
  - `defaultHandleRedirect`

### TODO post-merge
- [ ] Document new handlers
- [ ] Add tests for custom `handleMfaRequired` and `handlePkceRequired` methods

### CODE REVIEW

File: src/authentication.js [hunk 1]
```diff
@@ -8,90 +12,86 @@ export const authenticationData = {
...
-  if (!store.tenantId) {
-    console.warn(
-      "setFirstFactors: tried to set factors without a tenantId set."
-    );
-    return;
+export function setCookiesAndTokens(tokens) {
+  setCookie(tokens.access.value, tokens.access.cookieOptions, "access");
+  setCookie(tokens.id.value, tokens.id.cookieOptions, "id");
+  if (tokens.refresh && tokens.refresh.value) {
+    setCookie(tokens.refresh.value, tokens.refresh.cookieOptions, "refresh");
```
Comment (tyrw) [inline src/authentication.js:22]: Heads up that the diffs on this file don't make any sense.

`setCookiesAndTokens` and `handleLoginResponse` were added, and all other methods were moved to `mfa.js`

File: src/link.js [hunk 2]
```diff
@@ -33,36 +42,20 @@ export async function loginWithLink({ token, uuid, redirect } = {}) {
...
+    // Handle the API response to the login request
+    return handleLoginResponse({
+      data,
+      redirect,
+      handleUpstreamResponse,
+      handleMfaRequired,
+      handlePkceRequired,
+      handleTokens,
+      handleRedirect,
+    });
```
Comment (tyrw) [inline src/link.js:58]: This method was used to standardize handling of the API response across various signup & login methods.

File: test/authentication.spec.js [hunk 3]
```diff
@@ -19,226 +22,108 @@ const blankAuthenticationData = {
...
     }
   });
 
-  describe("setFirstFactors", () => {
-    it("should update the available first factors when passed a valid authentication object", async () => {
-      const authentication = {
-        firstFactors: [
-          {
-            channel: "email",
-            strategy: "password",
```
Comment (tyrw) [inline test/authentication.spec.js:33]: Moved most tests in this file to `mfa.spec.js`

**Review by RJFelix — APPROVED** [review]
Looks good!

For my own learning purposes - what's the context of these changes, what new behavior or feature do they support?

**Comment by tyrw** [comment]
> Looks good!
> 
> For my own learning purposes - what's the context of these changes, what new behavior or feature do they support?

Thanks! I'll add more docs, but essentially:

- `handleRedirect` and `handleTokens` will benefit React native, where the browser methods are not available 
- `handleUpstreamResponse` will help with passthroughs & migrations
- `handleMfaRequired` and `handlePkceRequired` will allow custom behavior in those circumstances, if desired

**Pull request merged**